src/main/scala/beam/agentsim/agents/vehicles/BeamVehicle.scala

```diff
@@ -47,160 +58,153 @@ trait VehicleData extends BeamAgentData with Vehicle {
+    if (actorRef.path.name.startsWith(ActorPrefixName)) {
+      Some(Id.create(actorRef.path.name.substring(ActorPrefixName.length), classOf[Vehicle]))
+    } else {
+      None
+    }
   }
-}
 
+  implicit def vehicleId2actorRef(vehicleId: Id[Vehicle])(implicit context: ActorContext, timeout: Timeout) = {
+    context.actorSelection(s"user/${buildActorName(vehicleId)}").resolveOne(timeout.duration)
+  }
+  case class GetVehicleLocationEvent(time: Double) extends org.matsim.api.core.v01.events.Event(time) {
+    override def getEventType: String = getClass.getName
+  }
 
-case class EnterVehicleTrigger(tick: Double, vehicleId: Id[Vehicle], driver: Option[ActorRef] = None, passengers: Option[List[ActorRef]] = None)  extends Trigger
+  case class AlightingConfirmation(vehicleId: Id[Vehicle])
+  case class BoardingConfirmation(vehicleId: Id[Vehicle])
 
-/**
-  * when somebody leaves the vehicle
-  * @param tick time when it heppens
-  * @param vehicleId id of vehicle
-  * @param driver driver who leaves vehicle
-  * @param passengers all passengers who leave this vehicle
-  */
-case class LeaveVehicleTrigger(tick: Double, vehicleId: Id[Vehicle], driver: Option[ActorRef] = None, passengers: Option[List[ActorRef]] = None)  extends Trigger
+  case class BecomeDriver(tick: Double, driver: Id[_], passengerSchedule: Option[PassengerSchedule] = None)
+  case class UnbecomeDriver(tick: Double, driver: Id[_])
+  case class BecomeDriverSuccess(passengerSchedule: Option[PassengerSchedule])
+  case class DriverAlreadyAssigned(vehicleId: Id[Vehicle], currentDriver: ActorRef)
+
+  case class EnterVehicle(tick: Double, passenger : Id[Vehicle])
+  case class ExitVehicle(tick: Double, passenger : Id[Vehicle])
+  case class VehicleFull(vehicleId: Id[Vehicle])
+
+  case class UpdateTrajectory(trajectory: Trajectory)
+}
 
-case class NoEnoughSeats(tick: Double, vehicleId: Id[Vehicle], passengers: List[ActorRef], requiredSeats: Int)  extends Trigger
 
-case class DriverAlreadyAssigned(tick: Double, vehicleId: Id[Vehicle], currentDriver: ActorRef)  extends Trigger
 
 /**
   * Defines common behavior for any vehicle. Communicate with PersonAgent
   * VehicleManager.
   * Passenger and driver can EnterVehicle and LeaveVehicle
   */
-trait BeamVehicle extends Resource with  BeamAgent[VehicleData] with TriggerShortcuts {
-
-  def driver: Option[ActorRef]
+trait BeamVehicle extends Resource with  BeamAgent[VehicleData] with TriggerShortcuts with HasServices{
+  override val id: Id[Vehicle]
+  def data: VehicleData
 
+  var driver: Option[ActorRef]
   /**
     * Other vehicle that carry this one. Like ferry or track may carry a car
-    *
-    * @return
     */
-  def carrier: Option[ActorRef]
-
-  def passengers: List[ActorRef]
+  var carrier: Option[ActorRef] = None
+  var passengers: ListBuffer[Id[Vehicle]] = ListBuffer()
+  var trajectory: Option[Trajectory] = None
 
-  def data: VehicleData
-
-  def trajectory: Trajectory
-
-  def powerTrain: Powertrain
```

Comment:
I would leave def instead of var in trait to have better immutability. Concrete implementation will manage state of trajectory internally by handling appropriated event